### PR TITLE
Take screenshots

### LINKED ISSUE #1757: Make 'When I take a screenshot' steps available for mobile executions
Please make steps:

When I take a screenshot
When I take a screenshot to '$path'

available for mobile.

### PR DESCRIPTION
**Pull request by ikalinin1**
closes #1757

### CODE REVIEW

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 1]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
 When I click on element located `By.id(a-z)`
 Then elements located `By.xpath(//span[@id='item-to-purchase'])` are sorted by text in ascending order
 ----
+
+=== Take a screenshot
+
+==== *_Info_*
+
+Takes a screenshot to a default screenshot folder defined by property `ui.screenshots.directory`.
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:405]:
```suggestion
Takes a screenshot and saves it to the default folder defined by the property `ui.screenshots.directory`.
```

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 2]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
...
+
+[TIP]
+Full path to a screenshot will be posted to logs and allure
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:408]:
```suggestion
TIP: The full path to the screenshot is posted to the logs and the report.
```

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 3]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
...
+
+[TIP]
+Screenshots directory should be defined relatively of project root
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:411]: is it really required? absolute path is not allowed?

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 4]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
...
+When I take screenshot
+----
+
+==== *_Usage_*
+
+.Take a screenshot
+[source,gherkin]
+----
+Scenario: Take a screenshot
+When I take screenshot
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:426]: it seems the example is not needed here

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 5]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
...
+----
+
+=== Take a screenshot to a path
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:429]:
```suggestion
=== Take a screenshot and save it to the folder
```

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 6]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
...
+
+==== *_Info_*
+
+Takes a screenshot to a path defined by user
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:433]:
```suggestion
Takes a screenshot and saves it to the default folder defined by user
```

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 7]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
...
+
+[TIP]
+Full path to a screenshot will be posted to logs and allure
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:436]:
```suggestion
TIP: The full path to the screenshot is posted to the logs and the report.
```

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 8]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
...
+
+==== *_Wording_*
+
+[source,gherkin]
+----
+When I take screenshot to path `$path`
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:442]:
```suggestion
When I take screenshot and save it to folder `$path`
```

File: docs/modules/plugins/partials/generic-ui-steps.adoc [hunk 9]
```diff
@@ -397,3 +397,60 @@ Given I am on a page with the URL 'https://grocery.by'
...
+----
+
+==== *_Parameters_*
+
+. `$path` - absolute path to a screenshot file
```
Comment (valfirst) [inline docs/modules/plugins/partials/generic-ui-steps.adoc:447]:
```suggestion
. `$path` - an absolute or relative path to the folder to save taken screenshot
```

File: vividus-extension-selenium/src/main/java/org/vividus/bdd/steps/ui/ScreenshotTakingSteps.java [hunk 10]
```diff
@@ -51,7 +51,7 @@ public void whenITakeScreenshot() throws IOException
      * @param path Path to the location for saving the screenshot
      * @throws IOException If an input or output exception occurred
      */
-    @When("I take a screenshot to '$path'")
+    @When("I take screenshot to `$path`")
     public void whenITakeScreenshotToPath(String path) throws IOException
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/bdd/steps/ui/ScreenshotTakingSteps.java:55]:
```suggestion
    public void whenITakeScreenshotToPath(Path path) throws IOException
```

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java [hunk 11]
```diff
@@ -45,6 +63,44 @@ public BufferedImage takeViewportScreenshot() throws IOException
         }
     }
 
+    protected Path takeScreenshotAsFile(Supplier<Path> pathProvider,
```
Comment (valfirst) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java:66]:
```suggestion
    protected Path takeScreenshotAsFile(Path screenshotPath,
```
?

File: vividus-tests/src/main/resources/data/tables/mobile_app/start.table [hunk 12]
```diff
@@ -0,0 +1,5 @@
+{valueSeparator=!}
+|step                                   |
+!Given I start mobile application with capabilities:!
+!|name|value     |!
+!|app |${app-url}|!
```
Comment (valfirst) [inline vividus-tests/src/main/resources/data/tables/mobile_app/start.table:5]: please format indents in the table to make it more readable

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java [hunk 13]
```diff
@@ -45,6 +63,42 @@ public BufferedImage takeViewportScreenshot() throws IOException
...
+        if (screenshotData.length > 0)
+        {
+            Path parent = screenshotFilePath.getParent();
+            if (parent != null)
+            {
+                FileUtils.forceMkdir(parent.toFile());
+            }
+            Files.write(screenshotFilePath, screenshotData);
+
+            LOGGER.info("Screenshot was taken: {}", screenshotFilePath.toAbsolutePath());
```
Comment (uarlouski) [inline vividus-extension-selenium/src/main/java/org/vividus/selenium/screenshot/AbstractScreenshotTaker.java:94]: use lazy logging

File: vividus-extension-selenium/src/main/resources/steps/defaults/screenshot.steps [hunk 14]
```diff
@@ -0,0 +1,5 @@
+Composite: When I take a screenshot
+When I take screenshot
+
+Composite: When I take a screenshot to '$path'
+When I take screenshot and save it to folder `<path>`
```
Comment (uarlouski) [inline vividus-extension-selenium/src/main/resources/steps/defaults/screenshot.steps:5]: missing eof

**Review by valfirst — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Pull request merged**